### PR TITLE
[BPF] Support signed division at cpu v1 with constant divisor

### PR DESCRIPTION
**Pull request by yonghong-song**
The motivation example likes below
```
  $ cat t1.c
  struct S {
    int var[3];
  };
  int foo1 (struct S *a, struct S *b)
  {
      return a - b;
  }
```
For cpu v1/v2/v3, the compilation will fail with the following errors:
```
  $ clang --target=bpf -O2 -c t1.c -mcpu=v3
  t1.c:4:5: error: unsupported signed division, please convert to unsigned div/mod.
      4 | int foo1 (struct S *a, struct S *b)
        |     ^
  1 error generated.
```
The reason is that sdiv/smod is supported at -mcpu=v4. At cpu v1/v2/v3, only udiv/umod is supported.

But the above example (for func foo1()) is reasonable common and user has to workaround the compilation failure by using udiv with conditionals.

For x86, for the above t1.c, compile and dump the asm code like below:
```
  $ clang -O2 -c t1.c && llvm-objdump -d t1.o
  0000000000000000 <foo1>:
       0: 48 29 f7                      subq    %rsi, %rdi
       3: 48 c1 ef 02                   shrq    $0x2, %rdi
       7: 69 c7 ab aa aa aa             imull   $0xaaaaaaab, %edi, %eax # imm = 0xAAAAAAAB
       d: c3                            retq
```
Basically sdiv can be replaced with sub, shr and imul. Latest gcc-bpf is also able to generate code similar to x86 with -mcpu=v1. See https://godbolt.org/z/feP9ETbjj

So let us add clang support for sdiv (constant divisor) as well at -mcpu=v1. But we still want to keep udiv untouched at -mcpu=v1. One more parameter "bool IsSigned" is added to isIntDivCheap(). The "IsSigned" parameter is used only by BPF backend to ensure udiv not impacted.

Note that only 32-bit sdiv (constant divisor) can be converted into sub/shr/imul. 64-bit sdiv (constant divisor) cannot be converted since bpf does not support 64-bit multiplication without potential overflow.

### CODE REVIEW

File: llvm/include/llvm/CodeGen/BasicTTIImpl.h [hunk 1]
```diff
@@ -565,7 +565,12 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
       if (!isa<ConstantInt>(Inst.getOperand(1)))
         return false;
       EVT VT = getTLI()->getValueType(DL, Inst.getType());
-      return !getTLI()->isIntDivCheap(VT, Fn.getAttributes());
+
+      bool IsSigned = true;
+      if (Inst.getOpcode() == Instruction::SDiv ||
+          Inst.getOpcode() == Instruction::SRem)
+        IsSigned = false;
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/BasicTTIImpl.h:572]: can just initialize isSigned with the compare result

File: llvm/lib/Target/BPF/BPFISelLowering.cpp [hunk 2]
```diff
@@ -238,6 +238,11 @@ bool BPFTargetLowering::isZExtFree(SDValue Val, EVT VT2) const {
   return TargetLoweringBase::isZExtFree(Val, VT2);
 }
 
+bool BPFTargetLowering::isIntDivCheap(EVT VT, bool IsSigned,
+                                      AttributeList Attr) const {
+  return (HasMovsx || !IsSigned) ? true : false;
```
Comment (arsenm) [inline llvm/lib/Target/BPF/BPFISelLowering.cpp:243]:
```suggestion
  return HasMovsx || !IsSigned;
```

File: llvm/include/llvm/CodeGen/TargetLowering.h [hunk 3]
```diff
@@ -549,7 +549,9 @@ class TargetLoweringBase {
   /// several shifts, adds, and multiplies for this target.
   /// The definition of "cheaper" may depend on whether we're optimizing
   /// for speed or for size.
-  virtual bool isIntDivCheap(EVT VT, AttributeList Attr) const { return false; }
+  virtual bool isIntDivCheap(EVT VT, bool IsSigned, AttributeList Attr) const {
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/TargetLowering.h:552]: Maybe just pass through the opcode directly than IsSigned (although then you need to decide between an Instruction opcode or ISD opcode)

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 4]
```diff
@@ -5377,11 +5377,13 @@ SDValue TargetLowering::SimplifySetCC(EVT VT, SDValue N0, SDValue N1,
       N0.hasOneUse() && (Cond == ISD::SETEQ || Cond == ISD::SETNE)) {
     // When division is cheap or optimizing for minimum size,
     // fall through to DIVREM creation by skipping this fold.
-    if (!isIntDivCheap(VT, Attr) && !Attr.hasFnAttr(Attribute::MinSize)) {
-      if (N0.getOpcode() == ISD::UREM) {
+    bool IsSigned = N0.getOpcode() == ISD::SREM;
+    if (!isIntDivCheap(VT, IsSigned, Attr) &&
+        !Attr.hasFnAttr(Attribute::MinSize)) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:5382]: Swap the order of checks

File: llvm/test/CodeGen/BPF/sdiv_to_mul.ll [hunk 5]
```diff
@@ -0,0 +1,80 @@
+; RUN: llc -O2 -march=bpfel -mcpu=v1 < %s | FileCheck --check-prefix=CHECK-V1 %s
+; RUN: llc -O2 -march=bpfel -mcpu=v3 < %s | FileCheck --check-prefix=CHECK-V3 %s
+; RUN: llc -O2 -march=bpfel -mcpu=v4 < %s | FileCheck --check-prefix=CHECK-V4 %s
```
Comment (arsenm) [inline llvm/test/CodeGen/BPF/sdiv_to_mul.ll:3]: Don't need -O2

File: llvm/test/CodeGen/BPF/sdiv_to_mul.ll [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+;     return a%15;
+;   }
+
+target triple = "bpf"
+
+; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
+define dso_local i32 @foo1(ptr noundef %a, ptr noundef %b) local_unnamed_addr #0 {
+entry:
+  %sub.ptr.lhs.cast = ptrtoint ptr %a to i64
+  %sub.ptr.rhs.cast = ptrtoint ptr %b to i64
```
Comment (arsenm) [inline llvm/test/CodeGen/BPF/sdiv_to_mul.ll:28]: Avoid the ptrtoint, and just make the arguments i64 to start with?

File: llvm/test/CodeGen/BPF/sdiv_to_mul.ll [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+;     return a - b;
+;   }
+;   int foo2(int a)
+;   {
+;     return a/15;
+;   }
+;   int foo3(int a)
+;   {
+;     return a%15;
+;   }
```
Comment (arsenm) [inline llvm/test/CodeGen/BPF/sdiv_to_mul.ll:20]: These cases are simple enough that they don't really need a source reference

File: llvm/test/CodeGen/BPF/sdiv_to_mul.ll [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+; CHECK-V4:       w[[#]] s%= 15
+
+attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) "frame-pointer"="all" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="v1" }
+
+!llvm.module.flags = !{!0, !1}
+!llvm.ident = !{!2}
+
+!0 = !{i32 1, !"wchar_size", i32 4}
+!1 = !{i32 7, !"frame-pointer", i32 2}
+!2 = !{!"clang version 20.0.0git (git@github.com:yonghong-song/llvm-project.git 238f3f994a96c511134ca1bc11d2d03e4368a0c1)"}
```
Comment (arsenm) [inline llvm/test/CodeGen/BPF/sdiv_to_mul.ll:80]: Drop metadata

File: llvm/test/CodeGen/BPF/sdiv_to_mul.ll [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+entry:
+  %rem = srem i32 %a, 15
+  ret i32 %rem
+}
+
+; CHECK-V1-NOT:   r[[#]] s%= 15
+; CHECK-V3-NOT:   w[[#]] s%= 15
+; CHECK-V4:       w[[#]] s%= 15
+
+attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) "frame-pointer"="all" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="v1" }
```
Comment (arsenm) [inline llvm/test/CodeGen/BPF/sdiv_to_mul.ll:73]: Drop unnecessary attributes

File: llvm/test/CodeGen/BPF/sdiv_to_mul.ll [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
```
Comment (arsenm) [inline llvm/test/CodeGen/BPF/sdiv_to_mul.ll:74]: Test vector cases too

File: llvm/include/llvm/CodeGen/BasicTTIImpl.h (hunk 1 above)
Comment (4ast) [reply llvm/include/llvm/CodeGen/BasicTTIImpl.h:572]: I don't think we need to differentiate signed vs unsigned here.
Both kinds of division are expensive. This transformation is necessary in all cases.
That should avoid adding isSigned everywhere.

**Review by 4ast — CHANGES_REQUESTED** [review]

**Comment by yonghong-song** [comment]
Okay, I will treat signed and unsigned div the same for BPF. This way, all non-BPF codes will be removed. Will close this one and create another pull request.
